### PR TITLE
Add a default fallback to build_constraint

### LINKED ISSUE #1555: Ambiguous error message for second-order cone constraint
When trying to create a model with a second-order cone constraint and wrong comma-notation for the vector `[3,x]`
```julia
  model = Model(with_optimizer(MosekOptimizer));
  @variable(model, x[1:2]);
  @constraint(model,  [3,x] in SecondOrderCone())
```
I get the following ambigious error message:
```
ERROR: MethodError: no method matching build_constraint(::getfield(JuMP, Symbol("#_error#72")){Symbol}, ::Array{Any,1}, ::MathOptInterface.SecondOrderCone)
Closest candidates are:
  build_constraint(::Function, ::Any, ::Any, ::Any) at /Users/x/.julia/packages/JuMP/LjMor/src/macros.jl:441
  build_constraint(::Function, ::AbstractArray{T,1} where T, ::JuMP.AbstractVectorSet) at /Users/x/.julia/packages/JuMP/LjMor/src/sets.jl:13
  build_constraint(::Function, ::Array{#s390,1} where #s390<:JuMP.AbstractJuMPScalar, ::MathOptInterface.AbstractVectorSet) at /Users/x/.julia/packages/JuMP/LjMor/src/macros.jl:411
  ...
Stacktrace:
 [1] build_constraint(::Function, ::Array{Any,1}, ::SecondOrderCone) at /Users/x/.julia/packages/JuMP/LjMor/src/sets.jl:13
 [2] top-level scope at /Users/x/.julia/packages/JuMP/LjMor/src/macros.jl:516
```
_(follow-up on the Gitter discussion)_

### PR DESCRIPTION
**Pull request by odow**
Closes #1555

The specific error that `x` should be splatted is best left for the documentation (e.g in https://github.com/JuliaOpt/JuMP.jl/pull/1597)

### CODE REVIEW

File: src/macros.jl [hunk 1]
```diff
@@ -390,6 +390,14 @@ function parse_constraint(_error::Function, args...)
           "       expr1 == expr2\n" * "       lb <= expr <= ub")
 end
 
+# Generic fallback.
+function build_constraint(_error::Function, func,
+        set::Union{MOI.AbstractScalarSet, MOI.AbstractVectorSet})
+    return _error("unable to add the constraint because something is wrong " *
```
Comment (mlubin) [inline src/macros.jl]: We can be a bit more specific. The issue is that we don't recognize `func` as a JuMP function.

File: test/constraint.jl [hunk 2]
```diff
@@ -177,6 +177,21 @@ function constraints_test(ModelType::Type{<:JuMP.AbstractModel})
...
+        @variable(model, x[1:2])
+        err = ErrorException(
+            "In @constraint(model,[3, x] in SecondOrderCone()): unable to add" *
+            " the constraint because we don't recognise $([3, x]) as a valid " *
+            "JuMP function."
+        )
+        if VERSION < v"0.7"
+            @test_throws ErrorException @constraint(model, [3, x] in SecondOrderCone())
+        else
+            @test_throws err @constraint(model, [3, x] in SecondOrderCone())
```
Comment (blegat) [inline test/constraint.jl]: I didn't know we could do that, neat!

File: src/macros.jl [hunk 3]
```diff
@@ -390,6 +390,14 @@ function parse_constraint(_error::Function, args...)
           "       expr1 == expr2\n" * "       lb <= expr <= ub")
 end
 
+# Generic fallback.
+function build_constraint(_error::Function, func,
+        set::Union{MOI.AbstractScalarSet, MOI.AbstractVectorSet})
+    return _error("unable to add the constraint because we don't recognise " *
```
Comment (mlubin) [inline src/macros.jl]: Nit: recognize

File: test/constraint.jl [hunk 4]
```diff
@@ -177,6 +177,21 @@ function constraints_test(ModelType::Type{<:JuMP.AbstractModel})
         # @test c.set == MOI.SecondOrderCone(1)
     end
 
+    @testset "Syntax error" begin
+        model = ModelType()
+        @variable(model, x[1:2])
+        err = ErrorException(
+            "In @constraint(model,[3, x] in SecondOrderCone()): unable to add" *
+            " the constraint because we don't recognise $([3, x]) as a valid " *
```
Comment (mlubin) [inline test/constraint.jl]: same here

**Comment by odow** [comment]
Updated spelling. Okay to merge from my end.

**Pull request merged**